src/libstd/collections/hash/raw/mod.rs

```diff
@@ -0,0 +1,1210 @@
+            ctrl,
+            bucket_mask: buckets - 1,
+            items: 0,
+            growth_left: bucket_mask_to_capacity(buckets - 1),
+        })
+    }
+
+    /// Attempts to allocate a new hash table with at least enough capacity
+    /// for inserting the given number of elements without reallocating.
+    fn try_with_capacity(
+        capacity: usize,
+        fallability: Fallibility,
+    ) -> Result<RawTable<T>, CollectionAllocErr> {
+        if capacity == 0 {
+            Ok(RawTable::new())
+        } else {
+            unsafe {
+                let buckets =
+                    capacity_to_buckets(capacity).ok_or_else(|| fallability.capacity_overflow())?;
+                let result = RawTable::new_uninitialized(buckets, fallability)?;
+                result
+                    .ctrl(0)
+                    .write_bytes(EMPTY, result.buckets() + Group::WIDTH);
+
+                // If we have fewer buckets than the group width then we need to
+                // fill in unused spaces in the trailing control bytes with
+                // DELETED entries. See the comments in set_ctrl.
+                if result.buckets() < Group::WIDTH {
+                    result
+                        .ctrl(result.buckets())
+                        .write_bytes(DELETED, Group::WIDTH - result.buckets());
+                }
+
+                Ok(result)
+            }
+        }
+    }
+
+    /// Allocates a new hash table with at least enough capacity for inserting
+    /// the given number of elements without reallocating.
+    pub fn with_capacity(capacity: usize) -> RawTable<T> {
+        RawTable::try_with_capacity(capacity, Fallibility::Infallible)
+            .unwrap_or_else(|_| unsafe { hint::unreachable_unchecked() })
+    }
+
+    /// Deallocates the table without dropping any entries.
+    #[inline]
+    unsafe fn free_buckets(&mut self) {
+        let (layout, _) =
+            calculate_layout::<T>(self.buckets()).unwrap_or_else(|| hint::unreachable_unchecked());
+        dealloc(self.ctrl.as_ptr(), layout);
+    }
+
+    /// Returns the index of a bucket from a `Bucket`.
+    #[inline]
+    unsafe fn bucket_index(&self, bucket: &Bucket<T>) -> usize {
+        offset_from(bucket.ptr.as_ptr(), self.data.as_ptr())
+    }
+
+    /// Returns a pointer to a control byte.
+    #[inline]
+    unsafe fn ctrl(&self, index: usize) -> *mut u8 {
+        debug_assert!(index < self.buckets() + Group::WIDTH);
+        self.ctrl.as_ptr().add(index)
+    }
+
+    /// Returns a pointer to an element in the table.
+    #[inline]
+    pub unsafe fn bucket(&self, index: usize) -> Bucket<T> {
+        debug_assert_ne!(self.bucket_mask, 0);
+        debug_assert!(index < self.buckets());
+        Bucket::from_ptr(self.data.as_ptr().add(index))
+    }
+
+    /// Erases an element from the table without dropping it.
+    #[inline]
+    pub unsafe fn erase_no_drop(&mut self, item: &Bucket<T>) {
+        let index = self.bucket_index(item);
```

Comment:
I don't think this makes a big difference in practice, the compiler can optimize `bucket_index` back to the original index if this function is inlined. I chose to return a pointer to the element because in most cases you are not deleting and will just be accessing the element.

Reply:
Actually I think there might be a legitimate bug here if the values are ZSTs. I will probably need some special handling for this case.